core/store/src/trie/shard_tries.rs

```diff
@@ -445,21 +446,35 @@ impl ShardTries {
     /// Loads in-memory tries upon startup. The given shard_uids are possible candidates to load,
     /// but which exact shards to load depends on configuration. This may only be called when flat
     /// storage is ready.
+    ///
+    /// The `extra_load_mem_tries_for_shards` parameter is used to load memtries
+    /// for shards that are not configured to be loaded but should be loaded
+    /// anyway. This is used when a shard is about to be resharded and we need
+    /// to have the memtries loaded for it.
     pub fn load_mem_tries_for_enabled_shards(
         &self,
         tracked_shards: &[ShardUId],
+        extra_load_mem_tries_for_shards: &HashSet<ShardUId>,
         parallelize: bool,
     ) -> Result<(), StorageError> {
         let trie_config = &self.0.trie_config;
         let shard_uids_to_load = tracked_shards
             .iter()
-            .copied()
             .filter(|shard_uid| {
                 trie_config.load_mem_tries_for_tracked_shards
                     || trie_config.load_mem_tries_for_shards.contains(shard_uid)
+                    || extra_load_mem_tries_for_shards.contains(shard_uid)
```

Comment:
I have a concern about it:
If the resharding is planned 2 epochs later, and we do not care about parent shard in this or next epoch, then the parent shard won't be contained in `tracked_shards` and won't be loaded?

Reply:
I mean, `tracked_shards` is already filtered with:
```
let tracked_shards: Vec<_> = shard_uids
            .iter()
            .filter(|shard_uid| {
                shard_tracker.care_about_shard(
```

Reply:
Hmm, agree with this. While this solution works for the case we are trying to handle, i.e. RPC nodes that track all shards, this isn't a general solution.

What if instead we call some async version of this function (like the state sync one) at the epoch boundary (first block of epoch)? @staffik, would that work?

Reply:
I worry loading memtrie for a currently active shard might not work out of the box.
In this case, I think something in the lines of `shard_uids.filter(extra_load_mem_tries_for_shards` would just work.

Sanity check: if we gonna track child shard but we do not track parent shard, we state sync to parent shard and do we load memtrie after state sync?

Reply:
From another point of view, we could assert that the node is configured to track all shard if `extra_load_mem_tries_for_shards` is used and/or non empty.

It is nice to have a simple logic even if the solution applies only to RPC nodes (it's tailored for them after all)

Reply:
Yeah good point but I think it should be covered by the state sync integration. 

case a
Let's say a validator node doesn't track shard S on startup. A few epochs later it is assigned shard S. I think at this point it should do the same check as here, whether S will be resharded within this client and if so load it into memory. 

case b
Let's say a validator node doesn't track shard S on startup. A few epochs later it is assign a child of shard S, called S'. Let's say it's the first epoch where S' exists. Here the node should get S in the current epoch, load it into memory and perform resharding to get S'. 

cc @marcelo-gonzalez thoughts? 

@staffik and @shreyan-gupta Are you happy to keep this PR as is and cover the other cases separately? I think some code may be reusable but we can just refactor it as part of the state sync work.

Reply:
Aah wait, one thing I don't understand, currently we don't enable memtries for RPC nodes right? Where is that handled in the codebase? I don't see any "if rpc" conditions in the `load_mem_tries_for_enabled_shards` or any other function.

Also, like Wac mentioned, for normal nodes that didn't track S or S', state sync should be able to handle it without issues and we don't need to do anything extra. This PR should probably only address RPC nodes (that track all shards)?

Reply:
> we don't enable memtries for RPC nodes right? Where is that handled in the codebase?

It is handled in `config.json` by RPC operator, by setting `load_mem_tries_for_tracked_shards: false`

Reply:
> Aah wait, one thing I don't understand, currently we don't enable memtries for RPC nodes right? Where is that handled in the codebase? I don't see any "if rpc" conditions in the load_mem_tries_for_enabled_shards or any other function.

I believe RPC can decide for itself what shards it wants to load by setting the appropriate configuration. 

I recall validators should be forced to enable `load for tracked` config, there should be an assertion somewhere, @staffik will know where it is.

Reply:
@wacban Did you mean this: https://github.com/near/nearcore/blob/master/chain/client/src/client_actor.rs#L330-L334 ?

Reply:
heh, I did mean that line but it doesn't do what I expected. This one is for single shard tracking. Do we not force validators to use memtrie today?

Reply:
I think we do not force them programatically, though we tell them to enable memtries, otherwise they are likely to fall behind after we changed gas cost or sth like that

Reply:
> cc @marcelo-gonzalez thoughts?
> 

Yeah I think that description is right. Case b is actually the motivation behind this PR: https://github.com/near/nearcore/pull/12404